include/forcing/ForcingsEngineDataProvider.hpp

```diff
@@ -0,0 +1,211 @@
+    {
+        auto pos = data_.find(key);
+        if (pos == data_.end()) {
+          return nullptr;
+        }
+
+        return pos->second;
+    }
+
+    void set(const key_type& key, value_type value)
+    {
+        data_[key] = value;
+    }
+
+    void clear()
+    {
+        data_.clear();
+    }
+
+  private:
+    //! Instance map of underlying BMI models.
+    std::unordered_map<key_type, value_type> data_;
+};
+
+} // namespace detail
+
+
+//! Forcings Engine Data Provider
+//! @tparam DataType Data type for values returned from derived classes
+//! @tparam SelectionType Selector type for querying data from derived classes
+template<typename DataType, typename SelectionType>
+struct ForcingsEngineDataProvider
+  : public DataProvider<DataType, SelectionType>
+{
+    using data_type = DataType;
+    using selection_type = SelectionType;
+    using clock_type = std::chrono::system_clock;
+
+    ~ForcingsEngineDataProvider() override = default;
+
+    boost::span<const std::string> get_available_variable_names() override
+    {
+        return var_output_names_;
+    }
+
+    long get_data_start_time() override
+    {
+        return clock_type::to_time_t(time_begin_);
+    }
+
+    long get_data_stop_time() override
+    {
+        return clock_type::to_time_t(time_end_);
+    }
+
+    long record_duration() override
+    {
+        return std::chrono::duration_cast<std::chrono::seconds>(time_step_).count();
+    }
+
+    size_t get_ts_index_for_time(const time_t& epoch_time) override
+    {
+        const auto epoch = clock_type::from_time_t(epoch_time);
+
+        if (epoch < time_begin_ || epoch > time_end_) {
+          throw std::out_of_range{
+            "epoch " + std::to_string(epoch.time_since_epoch().count()) +
+            " out of range of " + std::to_string(time_begin_.time_since_epoch().count()) + ", "
+            + std::to_string(time_end_.time_since_epoch().count())
+          };
+        }
+
+        return (epoch - time_begin_) / time_step_;
+    }
+
+    std::shared_ptr<models::bmi::Bmi_Py_Adapter> model() noexcept
+    {
+        return bmi_;
+    }
+
+    /* Remaining virtual member functions from DataProvider must be implemented
+       by derived classes. */
+
+    data_type get_value(
+      const selection_type& selector,
+      data_access::ReSampleMethod m
+    ) override = 0;
+
+    std::vector<data_type> get_values(
+      const selection_type& selector,
+      data_access::ReSampleMethod m
+    ) override = 0;
+
+  protected:
+    using storage_type = detail::ForcingsEngineStorage;
+
+    ForcingsEngineDataProvider(
+        std::string init,
+        std::size_t time_begin_seconds,
+        std::size_t time_end_seconds
+    )
+      : time_begin_(std::chrono::seconds{time_begin_seconds})
+      , time_end_(std::chrono::seconds{time_end_seconds})
+    {
+        // Get a forcings engine instance if it exists for this initialization file
+        auto bmi_ = storage_type::instances.get(init);
```

Comment:
As written, this shadows the member variable `bmi_`, and so doesn't actually initialize it. I think you wanted simply
```suggestion
        bmi_ = storage_type::instances.get(init);
```

Reply:
Good catch, yeah I missed that when I removed the factory function. Thanks!